<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">

<head>
<title>README.html</title>

</head>

<body>

<h1>miR-PREFeR: <strong>mi</strong>cro<strong>R</strong>NA <strong>PRE</strong>diction <strong>F</strong>rom small <strong>R</strong>NAseq data</h1>

<p><strong>The miR-PREFeR pipeline is still in active development, to be enable to use the newest features from the pipeline, it's better to check the  <a href="https://github.com/hangelwen/miR-PREFeR">https://github.com/hangelwen/miR-PREFeR</a> page and obtain the newest version. The current version is only tested under Python 2.6.7, Python 2.7.2 and Python 2.7.3 and should work under Python 2.6.* and Python 2.7.*. It does NOT work under Python 3.0 currently</strong></p>

<h2>1. Required programs</h2>

<p>To run the miR-PREFeR pipeline, the ViennaRNA package(version 1.8.5 or 2.1.x) and samtools(0.1.15 or later. Tested under 0.1.18) should be installed on the system. miR-PREFeR uses samtools commands to manipulate SAM and BAM alignment files, and uses RNALfold from the ViennaRNA package to do RNA secondary structure folding. The miR-PREFeR pipeline takes SAM alignment files as input, so an aligner such as Bowtie is also required (Not required by the pipeline, but needed for preparing the input data).</p>

<h3>ViennaRNA package</h3>

<p>The ViennaRNA package can be downloaded from <a href="http://www.tbi.univie.ac.at/~ronny/RNA/index.html.">http://www.tbi.univie.ac.at/~ronny/RNA/index.html.</a>. The website provides precompiled  packages for some platforms. For these platforms, download the corresponding package and install it according to the platform package management system. If no precompiled package is provided for your platform, you can install it from source code.
To install the package from source code, simply decompress the source code archive and cd into the folder,
then:</p>

<pre><code>./configure
make
</code></pre>

<p>and (as root):
    make install</p>

<p>If the user does not have root permission. then the package can be installed to
user specified locations:</p>

<pre><code>./configure --prefix="/path/to/install" --without-perl
make
make install
</code></pre>

<p>After finish installing the package, <strong>add the bin directory to the PATH environment variable.</strong> Instructions on how to add a path to the PATH environment variable can be found at <a href="http://www.cyberciti.biz/faq/unix-linux-adding-path/">http://www.cyberciti.biz/faq/unix-linux-adding-path/</a> or just google "<code>add directory to PATH</code>".</p>

<p><strong>NOTE: Because that RNALfold from the ViennaRNA package version 2.0.4 has a bug (If the input sequence has no valid folding structure, the program produces a segmentation fault), please make sure to use ViennaRNA package version 1.8.x or the newest 2.1.2.</strong> The 1.8.* version is easier to install than the 2.1.2 version. Because the new version needs the <code>gengetopt</code> package (<a href="http://www.gnu.org/s/gengetopt/gengetopt.html">http://www.gnu.org/s/gengetopt/gengetopt.html</a>), which may not be available on your system.</p>

<h3>Samtools</h3>

<p>Samtools can be downloaded from <a href="http://samtools.sourceforge.net/">http://samtools.sourceforge.net/</a>. Please follow the instructions from the package to install it. Please note the version 0.1.15 or later is needed (The pipeline uses the <code>samtools depth</code> command, which was introduced from version 0.1.15).</p>

<h2>2. Obtain and install the pipeline</h2>

<p>The miR-PREFeR pipeline is hosted on Github (https://github.com/hangelwen/miR-PREFeR). To obtain the program, open an terminal window and execute the following command:</p>

<pre><code>git clone https://github.com/hangelwen/miR-PREFeR.git
</code></pre>

<p>This makes an directory named 'miR-PREFeR' in the current directory, and clones the newest version of the miR-PREFeR pipeline code into the directory. The pipeline is ready to use.</p>

<p>If you do not have <code>git</code> on your system, simply go to <a href="https://github.com/hangelwen/miR-PREFeR">https://github.com/hangelwen/miR-PREFeR</a>, download a zip file by clicking 'Download ZIP', and unzip the file.</p>

<p><strong>The miR-PREFeR pipeline is still in active development, to be enable to use the newest features from the pipeline, it's better to check the  <a href="https://github.com/hangelwen/miR-PREFeR">https://github.com/hangelwen/miR-PREFeR</a> page and obtain the newest version.</strong></p>

<h2>3. How to run the pipeline.</h2>

<h3>a. Prepare input data for the pipeline.</h3>

<p>The miR-PREFeR pipeline needs the following input:</p>

<ol>
<li>A fasta file, which contains the gnome sequences of the species under study.</li>
<li>one or more SAM files which contains the alignments of small RNAseq data with the gnome.</li>
<li>(Optional) An GFF (<a href="http://www.sanger.ac.uk/resources/software/gff/spec.html">http://www.sanger.ac.uk/resources/software/gff/spec.html</a>) file which lists regions in the gnome sequences that should be ignored from miRNA analysis.</li>
</ol>

<h4>a). Genome fasta file</h4>

<p>Fasta format specification can be found at <a href="http://www.ncbi.nlm.nih.gov/BLAST/blastcgihelp.shtml">http://www.ncbi.nlm.nih.gov/BLAST/blastcgihelp.shtml</a>. In miR-PREFeR, when an identifier of a genomic sequence is needed, the first part that does not contain any white space characters (whitespace, tab, etc) is used. For example, for the following sequence, 'ath-MIR773a' is used as the identifier of the seqeunce. <strong>Thus, please ensure that all the sequences in the FASTA files have different identifiers.</strong></p>

<pre><code>&gt;ath-MIR773a MI0005103
AGGAGGCAAUAGCUUGAGCAAAUAAUUGAUUGCAGAAGUCCAUCGACUAAAGCUGUCACCUGUUUGCUUCCAGCUUUUGUCUCCU
</code></pre>

<h4>b). SAM alignment files</h4>

<p>The miR-PREFeR pipeline takes SAM format alignment files. SAM alignment files can be generated by many aligners. Here we use Bowtie (<a href="http://bowtie-bio.sourceforge.net/index.shtml">http://bowtie-bio.sourceforge.net/index.shtml</a>) as an example.</p>

<p><strong>Prepare RNAseq fasta files for Bowtie</strong></p>

<p>Because miR-PREFeR is able to use multiple small RNAseq data as input and utilizes information from multiple RNAseq samples (expression pattern consistency, differential expression, etc) to increase the performance of the miRNA loci prediction. <strong>Thus, It's an requirement that the RNAseq reads data (fasta files that contain the RNAseq reads) be preprocessed by the provided <code>process-reads-fasta.py</code> script.</strong> The script takes an list of <strong>uncollapsed fasta files</strong> as input, and renames the reads in the input files, and produces a list of collapsed fasta files. Here <code>uncollpased fasta</code> means that identical reads in the fasta file have multiple entries.</p>

<p>For example, if you have 3 small RNAseq samples in uncollpased fasta format with names <code>SAMPLE1.fasta</code>, <code>SAMPLE2.fasta</code>, and <code>SAMPLE3.fasta</code>, then by executing:</p>

<pre><code>python process-reads-fasta.py SAMPLE1.fasta SAMPLE2.fasta SAMPLE3.fasta
</code></pre>

<p>Three new files with name <code>SAMPLE1.fasta.processed</code>, <code>SAMPLE2.fasta.processed</code>, and <code>SAMPLE3.fasta.processed</code> are produced. These three files are then aligned to the genome fasta file using Bowtie.</p>

<p><strong>Align the RNA-seq fasta files with Bowtie:</strong></p>

<p>a1. indexing the genome sequences:</p>

<pre><code>bowtie-build -q -f TAIR10.fas bowtie-index/TAIR10
</code></pre>

<p>a2. alignment</p>

<pre><code>bowtie -a -v 0 -p 8 -m 10 -S bowtie-index/TAIR10  -f SAMPLE1.fasta.processed &gt; SAMPLE1.sam 2&gt; SMAPLE1.log
bowtie -a -v 0 -p 8 -m 10 -S bowtie-index/TAIR10  -f SAMPLE2.fasta.processed &gt; SAMPLE2.sam 2&gt; SMAPLE2.log
bowtie -a -v 0 -p 8 -m 10 -S bowtie-index/TAIR10  -f SAMPLE3.fasta.processed &gt; SAMPLE3.sam 2&gt; SMAPLE3.log
</code></pre>

<p>The -v 0 option allows no mismatch in the alignment
The -m 10 option discards reads that can be mapped to more than 10 positions.
The -p 8 option uses 8 threads to do the alignment
The -v 0 and the -m 10 options are recommended to use the values above.</p>

<h4>c). The (optional) gff file</h4>

<p>For a genome that already has known annotations, some regions in the genome can be ignored when doing the miRNA analysis. For example, many species have protein coding sequence (CDS) annotations, so there is no need to do run miR-PREFeR on those regions. These regions should be listed in a GFF file (<a href="http://www.sanger.ac.uk/resources/software/gff/spec.html">http://www.sanger.ac.uk/resources/software/gff/spec.html</a>).</p>

<h3>b. Prepare a configuration file for the pipeline.</h3>

<p>The miR-PREFeR.py script takes a configuration file as input. The configuration file lists all the information (such as where are the SAM files, genome fasta file, etc) needed to run the pipeline. An example configuration file (<code>example/config.example</code>) with detailed descriptions of each option is in the <code>example</code> folder. The options are also explained in the following section.</p>

<ol>
<li>PIPELINE_PATH: The path of the miR-PREFeR pipeline.</li>
<li>FASTA_FILE: The path of the genome sequences in fasta format. Absolute path preferred.</li>
<li>ALIGNMENT_FILE: The path of the RNAseq alignment files in SAM format. Absolute path preferred.</li>
<li>GFF_FILE: The path of the optional GFF file. Absolute path preferred.</li>
<li>PRECURSOR_LEN: The max length of a miRNA precursor. The default is 300</li>
<li>READS_DEPTH_CUTOFF: The first step of the pipeline is to identify candidate regions of the miRNA loci. If READS_DEPTH_CUTOFF = N, then genomic position that the mapped depth is smaller than N is not considered. Default value is 20.</li>
<li>NUM_OF_CORE: Number of CPUS/Cores avalible for this computation. If commented out or leave blank, 1 is used.</li>
<li>OUTFOLDER: Outputfolder. If not specified, use the current working directory.</li>
<li>NAME_PREFIX: Prefix for naming the output files. For portability, please DO NOT contain any spaces and special characters. The prefered includes 'A-Z', 'a-z', '0-9', and underscore '_'.</li>
<li>MAX_GAP: Maximum gap length between two contigs to form a candidate region. Default value is 100.</li>
</ol>

<h3>c. Run the pipeline.</h3>

<p>With the configuration file ready, the pipeline can be used to predict miRNA loci. The miR-PREFeR pipeline can be run as following:</p>

<pre><code>python miR-PREFeR.py command configfile
</code></pre>

<p>Where <code>command</code> could be one of <code>check</code>, <code>pipeline</code>, <code>prepare</code>, <code>candidate</code>, <code>fold</code>, and <code>recover</code>. (Run <code>python miR-PREFeR.py -h</code> to see help.)</p>

<ol>
<li><strong>check:</strong> Check the presence of the depended programs (RNALfold and samtools), and check the recovery information (Shows which stage the previous computation on the same data was ceased. See the <code>recover</code> command in 6.).</li>
<li><strong>pipeline:</strong> Run the whole pipeline. That is, run <code>prepare</code>, <code>candidate</code>, and <code>fold</code> sequentially. This is the normal way to run miR-PREFeR.</li>
<li><strong>prepare:</strong>  Run the first step of the pipeline. This step prepares some data files needed in the following steps.</li>
<li><strong>candidate:</strong>  Identify possible candidate regions. This step can ONLY be run if the <code>prepare</code> step has been finished on the configfile file.</li>
<li><strong>fold:</strong> Fold the candidate regions and predict miRNA loci. This step can ONLY be run if the <code>prepare</code> and <code>candidate</code> steps have been finished on the configfile file.</li>
<li><p><strong>recover:</strong> Tries its best to recover an unfinished job and continues to run from the unfinished step it was ceased. This is designed to easily continue a job other than re-run the whole pipeline from start. For example, if a job was kill half way (This happens sometime. For example, the job takes too long time and the user killed it. Or, if the job was run on an Cluster and was killed halfway because some resources exceed the max values.) In these cases, one does not need to run the pipeline from start, the job can be continued by using the <code>recover</code> command:</p>

<p><code>python miR-PREFeR.py recover configfile</code></p></li>
</ol>

<p><strong>NOTE on job recovery:</strong> To recover and continue a job, the intermediate output files in the temporary folder (See the Output section) should not be removed. All files needed to do recovery are in this folder.</p>

<h2>4. Output</h2>

<p>The predicted miRNA loci are written to a GFF file named NAME_PREFIX + "_miRNA.gff3" in the output folder, where 'NAME_PREFIX' is the prefix for naming output files which is specified in the configuration file.</p>

<p>There is a folder named NAME_PREFIX + "_tmp" in the output folder. The folder contains intermediate files generated by the pipeline. If the job is finished without interruption, the folder can be deleted. <strong>If the job is interrupted half way, and one decides to run the job in the <code>recover</code> mode to continue the computation, then the folder SHOULD NOT be removed. All files needed to do recovery are in this folder.</strong></p>

</body>
</html>
