<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">

<head>
<title>README.html</title>

</head>

<body>

<h1>miR-PREFeR: <strong>mi</strong>cro<strong>R</strong>NA <strong>PRE</strong>diction <strong>F</strong>rom small <strong>R</strong>NAseq data</h1>

<p><strong>Plant microRNA prediction tools that utilize small RNA sequencing data are emerging quickly. These existing tools have at least one of the following problems: 1. high false positive rate; 2. long running time; 3. work only for genomes in their databases; 4. hard to install or use. We develop miR-PREFeR, which uses expression patterns of miRNA and follows the criteria for plant microRNA annotation to accurately predict plant miRNAs from one or more small RNA-Seq data samples of the same species. We tested miR-PREFeR on several plant species. The results show that miR-PREFeR is sensitive, accurate, fast, and has low memory footprint.</strong></p>

<p><strong>The miR-PREFeR paper has been accepted by Bioinformatics. <a href="http://bioinformatics.oxfordjournals.org/content/early/2014/06/14/bioinformatics.btu380.abstract">http://bioinformatics.oxfordjournals.org/content/early/2014/06/14/bioinformatics.btu380.abstract</a></strong></p>

<p><strong>The miR-PREFeR pipeline is still under active development. To use the newest features from the pipeline, check the  <a href="https://github.com/hangelwen/miR-PREFeR">https://github.com/hangelwen/miR-PREFeR</a> page and obtain the newest version. The current version is only tested under Python 2.6.7, Python 2.7.2 and Python 2.7.3 and should work under Python 2.6.* and Python 2.7.* (Tested platforms: Linux, Mac OS). It does NOT work under Python 3.0 currently. If you find any problem, please contact the author at hangelwen@gmail.com.</strong></p>

<h2>1. Required programs</h2>

<p>To run the miR-PREFeR pipeline, the ViennaRNA package(tested under 1.8.5 or 2.1.2, 2.1.5) and samtools(0.1.15 or later. Tested under 0.1.18, 0.1.19) should be installed on the system. miR-PREFeR uses samtools commands to manipulate SAM and BAM alignment files, and uses RNALfold from the ViennaRNA package to do RNA secondary structure folding. The miR-PREFeR pipeline takes SAM alignment files as input, so an aligner such as Bowtie is also required (Not required by the pipeline, but needed for preparing the input data).</p>

<h3>ViennaRNA package</h3>

<p>The ViennaRNA package can be downloaded from <a href="http://www.tbi.univie.ac.at/~ronny/RNA/index.html.">http://www.tbi.univie.ac.at/~ronny/RNA/index.html.</a>. The website provides precompiled  packages for most popular Linux distributions. For these platforms, download the corresponding package and install it according to the platform package management system. If no precompiled package is provided for your platform, you can install it from source code.
To install the package from source code, simply decompress the source code (e.g. type "tar xvf *.tar.gz) archive and cd into the folder,
then:</p>

<pre><code>./configure
make
make install
</code></pre>

<p>The above commands only work if the user has the root permission.</p>

<p>If the user does not have root permission. then the package can be installed to
user specified locations (here I assume you install the program in directory /user/tools/ViennaRNA/):</p>

<pre><code>./configure --prefix="/user/tools/ViennaRNA/" --without-perl
make
make install
</code></pre>

<p>After finish installing the package, <strong>add the /user/tools/ViennaRNA/bin directory to the PATH environment variable.</strong> Instructions on how to add a path to the PATH environment variable can be found at <a href="http://www.cyberciti.biz/faq/unix-linux-adding-path/">http://www.cyberciti.biz/faq/unix-linux-adding-path/</a> or just google "<code>add directory to PATH</code>".</p>

<p><strong>NOTE: Because that RNALfold from the ViennaRNA package version 2.0.4 has a bug (If the input sequence has no valid folding structure, the program produces a segmentation fault), please make sure to use the newest version of the ViennaRNA package.</strong></p>

<h3>Samtools</h3>

<p>Samtools can be downloaded from <a href="http://samtools.sourceforge.net/">http://samtools.sourceforge.net/</a>. Please follow the instructions from the package to install it. Please note the version 0.1.15 or later is needed (The pipeline uses the <code>samtools depth</code> command, which was introduced from version 0.1.15). One example set of commands to install samtools are given below. First, suppose you downloaded samtools-0.1.19.tar.bz2 and saved it in a folder with address /Users/xyz/. Then:</p>

<pre><code>cd /Users/xyz/
tar jfx samtools-0.1.19.tar.bz2
cd samtools-0.1.19
make
</code></pre>

<p>If successful, you should see an executable program named "samtools" in the same folder. You need to add the samtools executable to the PATH environment variable.</p>

<h3>For Mac OS users</h3>

<h4>a. Homebrew users</h4>

<p>Homebrew  <a href="https://github.com/Homebrew/homebrew">https://github.com/Homebrew/homebrew</a> is a fantastic package manager for MacOS. If you use Homebrew to manage your package on your Mac, mir-prefer and all it's dependencies can be installed simply by the following command</p>

<pre><code>brew install mir-prefer
</code></pre>

<p>Instructions on how to install and use homebrew can be found at <a href="http://coolestguidesontheplanet.com/setting-up-os-x-mavericks-and-homebrew/">http://coolestguidesontheplanet.com/setting-up-os-x-mavericks-and-homebrew/</a>.</p>

<h4>b. Other Mac  users</h4>

<p>Since it maybe difficult for some Mac OS users to successfully compile C/C++ programs, we provide several pre-compiled versions of samtools and RNALfold, which are put in the dependency/Mac/ folder. Mac OS users can first try to use the pre-compiled programs. If they do not work (This is possible), then try to compile from the source code by following the instructions in this section. To test whether a pre-compiled program works on your machine, change directory to the folder that contains the program, and then:</p>

<pre><code>chmod +x RNALfold
./RNALfold -V
</code></pre>

<p>Here I used the pre-compiled RNALfold as an example. If you can see usage information from the output, then the program works on your machine. Then add the path of the RNALfold to the PATH environment variable, and RNALfold is correctly configured.</p>

<p>Note that in the dependency/Mac/ folder there are several folders named after specific Mac OS X version. If you can find the same version as your own system, then use pre-compiled programs under that folder. Otherwise, try to use programs which were compiled with lower Mac OS X version number.</p>

<h3>For Windows users</h3>

<p>The pipeline depends on Samtools, and a few Linux/Unix utilities such as <code>sort</code>. Windows users need to install a Linux environment emulator such as Cygwin (<a href="http://cygwin.com/">http://cygwin.com/</a>), then install Samtools. Please follow instructions on the Cygwin website (<a href="http://cygwin.com/install.html">http://cygwin.com/install.html</a>) for more information.</p>

<h2>2. Obtain and install the pipeline</h2>

<p>The miR-PREFeR pipeline is hosted on Github (https://github.com/hangelwen/miR-PREFeR). To obtain the program, open an terminal window and execute the following command:</p>

<pre><code>git clone https://github.com/hangelwen/miR-PREFeR.git
</code></pre>

<p>This makes an directory named 'miR-PREFeR' in the current directory, and clones the newest version of the miR-PREFeR pipeline code into the directory. The pipeline is ready to use.</p>

<p>If you do not have <code>git</code> on your system, simply go to <a href="https://github.com/hangelwen/miR-PREFeR/releases">https://github.com/hangelwen/miR-PREFeR/releases</a>, download the zip file or the tar.gz file of the latest version.</p>

<p><strong>The miR-PREFeR pipeline is still under active development, to use the newest features from the pipeline, check <a href="https://github.com/hangelwen/miR-PREFeR">https://github.com/hangelwen/miR-PREFeR</a> and obtain the newest version.</strong></p>

<h2>3. Test the pipeline.</h2>

<p>The package contains a small dataset for testing whether the pipeline works after you download it. Please refer to <a href="HOW_TO_RUN_EXAMPLE.txt">HOW_TO_RUN_EXAMPLE.txt</a> file to see more infomation about testing.</p>

<h2>4. How to run the pipeline.</h2>

<h3>a. Prepare input data for the pipeline.</h3>

<p>The miR-PREFeR pipeline needs the following input:</p>

<ol>
<li>A fasta file, which contains the gnome sequences of the species under study.</li>
<li>one or more SAM files which contains the alignments of small RNAseq data with the gnome.</li>
<li>(Optional) An GFF (<a href="http://www.sanger.ac.uk/resources/software/gff/spec.html">http://www.sanger.ac.uk/resources/software/gff/spec.html</a>) file which lists regions in the gnome sequences that should be ignored from miRNA analysis.</li>
</ol>

<h4>a). Genome fasta file</h4>

<p>Fasta format specification can be found at <a href="http://www.ncbi.nlm.nih.gov/BLAST/blastcgihelp.shtml">http://www.ncbi.nlm.nih.gov/BLAST/blastcgihelp.shtml</a>. In miR-PREFeR, for the string following ">", only the first word that  is delimited by any white space characters (whitespace, tab, etc) is used. For example, for the following sequence, 'ath-MIR773a' is used as the identifier of the seqeunce. <strong>Thus, please ensure that all the sequences in the FASTA files have different identifiers.</strong></p>

<pre><code>&gt;ath-MIR773a MI0005103
AGGAGGCAAUAGCUUGAGCAAAUAAUUGAUUGCAGAAGUCCAUCGACUAAAGCUGUCACCUGUUUGCUUCCAGCUUUUGUCUCCU
</code></pre>

<h4>b). SAM alignment files</h4>

<p>The miR-PREFeR pipeline takes SAM format alignment files. SAM alignment files can be generated by many aligners. Here we use Bowtie (<a href="http://bowtie-bio.sourceforge.net/index.shtml">http://bowtie-bio.sourceforge.net/index.shtml</a>) as an example.</p>

<p><strong>Step 1. Prepare RNAseq fasta files for Bowtie</strong></p>

<p><strong><em>a. Prepare RNAseq fasta files from uncollpased fasta files.</em></strong></p>

<p>Because miR-PREFeR is able to use multiple small RNAseq data as input and utilizes information from multiple RNAseq samples to increase the performance of the miRNA loci prediction. <strong>Thus, It's an requirement that the RNAseq reads data (fasta files that contain the RNAseq reads) be preprocessed by the provided <code>process-reads-fasta.py</code> script.</strong> The script is under the <code>scripts</code> folder. The script takes a file that contains a list of sample/library names and an list of <strong>uncollapsed fasta files</strong> as input, and renames the reads in the input files, and produces a list of collapsed fasta files. Here <code>uncollpased fasta</code> means that identical reads in the fasta file have multiple entries.</p>

<p>For example, if you have 3 small RNAseq samples in uncollpased fasta format with names <code>SAMPLE1.fasta</code>, <code>SAMPLE2.fasta</code>, and <code>SAMPLE3.fasta</code>, and the samples are from 'root', 'flower', and 'salt', respectively, then first create a file with the following three lines:</p>

<pre><code>root
flower
salt
</code></pre>

<p>If you name the file as <code>samplename.txt</code>, then run (assume you are in the directory where you installed miR-PREFeR. Otherwise, using the correct path to the process-reads-fasta.py script.):</p>

<pre><code>python scripts/process-reads-fasta.py samplename.txt SAMPLE1.fasta SAMPLE2.fasta SAMPLE3.fasta
</code></pre>

<p>Three new files with name <code>SAMPLE1.fasta.processed</code>, <code>SAMPLE2.fasta.processed</code>, and <code>SAMPLE3.fasta.processed</code> are produced <strong>in the same folder as the input fasta files</strong>. These three files are then aligned to the genome fasta file using Bowtie.</p>

<p>The following is a section of the processed fasta file from the root sample. The identifier of each read contains the sample name as the first part, then a sequential number rx as to identify the read, the last part is xN, where N is the depth (the number of occurrence) of the read. <strong>Note that this is the required format of the identifiers of the reads. The read identifier in the SAM alignment file MUST follow the format. Otherwise, the pipeline stops at the very early stage and prints message complaining the format of the SAM files.</strong></p>

<pre><code>&gt;root_r0_x1
ACTACTGCAAGGGCTGGCTCAACCCGC
&gt;root_r1_x5
TGGTTGCTGTCGCTGGTCGCTGGT
&gt;root_r2_x1
CAAGGACAACAAGTGACGCCG
&gt;root_r3_x186
ATAACCGTAGTAATTCTAG
&gt;root_r4_x1
CCCATATTTTCTCTGAGCCTT
&gt;root_r5_x1
ATACGGTTCGTTCTGG
&gt;root_r6_x31
AACTGCGAATGGCTCATTAAAT
&gt;root_r7_x1
AAAGGTCGACGCGGGCTCTGCCCG
&gt;root_r8_x1
GATCCGGTGAAGTGTTCGGATC
&gt;root_r9_x1
GGTAGAGATCGGAGG
&gt;root_r10_x24
GTGGTTGTAGTATAGCGGTTAG
&gt;root_r11_x2
TTTGAACGCAAGTTGCGCCCCAA
</code></pre>

<p><strong><em>b. Prepare RNAseq fasta files from mirdeep/mirdeep2 format collapsed fasta files.</em></strong></p>

<p>If you have fasta files in mirDeep, mirDeep2 format (Reads are already collapsed and the identifiers follow format "Identifer_xN", where 'Identifier' is an arbitrary string without any blank characters. 'N' is the number of occurrence (depth) of the read in the library.), we provide a script <code>convert-mirdeep2-fasta.py</code>, which is also under the <code>scripts</code> folder, to convert them to the required format for miR-PREFeR:</p>

<pre><code>python scripts/convert-mirdeep2-fasta.py samplename.txt SAMPLE1.mirdeep.fasta SAMPLE2.mirdeep.fasta SAMPLE3.mirdeep.fasta
</code></pre>

<p>Three new files with name <code>SAMPLE1.mirdeep.fasta.processed</code>, <code>SAMPLE2.mirdeep.fasta.processed</code>, and <code>SAMPLE3.mirdeep.fasta.processed</code> are produced <strong>in the same folder as the input fasta files</strong>. These three files are then aligned to the genome fasta file using Bowtie.</p>

<p><strong>Step 2. Align the RNA-seq fasta files with Bowtie:</strong></p>

<p>A script, <code>bowtie-align-reads.py</code>, is provided to align the read files in fasta format generated by <code>process-reads-fasta.py</code>, <code>convert-mirdeep2-fasta.py</code>, or <code>convert-readcount-file.py</code>. The script is under the <code>scripts</code> folder. Here is an example to align the processed fasta files generated in the previous stage using the script:</p>

<pre><code>python bowtie-align-reads.py -f -r TAIR10.fas -t bowtie-index/ -p 8 SAMPLE1.fasta.processed SAMPLE2.fasta.processed SAMPLE3.fasta.processed
</code></pre>

<p>The command first generates the bowtie index files in the bowtie-index/ folder for the TAIR10.fas reference sequences, and then uses 8 threads to align <code>SAMPLE1.fasta.processed</code>, <code>SAMPLE2.fasta.processed</code>, and <code>SAMPLE3.fasta.processed</code>. Unmapped alignments (bowtie outputs SAM files with both mappped and unmapped alignments.) are filtered using SAMtools (the -f opiton). The result files are in the same folder as the input files, with suffix ".sam". For more options available in <code>bowtie-align-reads.py</code>, please refer to the <a href="scripts/README-SCRIPTS.txt">scripts/README-SCRIPTS.txt</a> file, or run <code>bowtie-align-reads.py</code> with the <code>-h</code> option:</p>

<pre><code>python bowtie-align-reads.py -h
</code></pre>

<p>For those who prefer to use bowtie directly (for example, users want to more bowtie options), please refer to Section <a href="#5-conducting-alignment-using-bowtie">Conducting alignment using Bowtie</a>. The section contains the steps to do alignments using bowtie and the format of the SAM files required by the miR-PREFeR pipeline.</p>

<h4>c). The (optional) genome annotation file in gff3 format</h4>

<p>For a genome with known annotations, some regions in the genome can be excluded when doing the miRNA analysis. For example, many species have protein coding sequence (CDS) annotations, so there is no need to run miR-PREFeR on those regions. Most plant genomes contain highly repetitive regions such as transposons, these regions should also be excluded. These regions should be listed in a GFF file (<a href="http://www.sanger.ac.uk/resources/software/gff/spec.html">http://www.sanger.ac.uk/resources/software/gff/spec.html</a>). Note that for the gff file, you should not contain introns, because miRNAs could be in introns.</p>

<h3>b. Prepare a configuration file for the pipeline.</h3>

<p>The miR_PREFeR.py script takes a configuration file as input. The configuration file lists all the information (such as the location of  the SAM files, genome fasta file, etc) needed to run the pipeline. An example configuration file with detailed descriptions of each option is in the example data. The options are also explained in the following section.</p>

<ol>
<li>PIPELINE_PATH: The path of the miR-PREFeR pipeline.</li>
<li>FASTA_FILE: The path of the genome sequences in fasta format. Absolute path preferred.</li>
<li>ALIGNMENT_FILE: The path of the RNAseq alignment files in SAM format. Absolute path preferred.</li>
<li>GFF_FILE: The path of the optional GFF file. Absolute path preferred.</li>
<li>PRECURSOR_LEN: The max length of a miRNA precursor. The default is 300</li>
<li>READS_DEPTH_CUTOFF: The first step of the pipeline is to identify candidate regions of the miRNA loci. If READS_DEPTH_CUTOFF = N, then genomic position that the mapped depth is smaller than N is not considered. Default value is 10.</li>
<li>NUM_OF_CORE: Number of CPUS/Cores avalible for this computation. If commented out or leave blank, 1 is used.</li>
<li>OUTFOLDER: Outputfolder. If not specified, use the current working directory.</li>
<li>NAME_PREFIX: Prefix for naming the output files. For portability, please DO NOT contain any spaces and special characters. The prefered includes 'A-Z', 'a-z', '0-9', and underscore '_'.</li>
<li>MAX_GAP: Maximum gap length between two contigs to form a candidate region. Default value is 100.</li>
<li>CHECKPOINT_SIZE: The pipeline makes a checkpoint after each major step. In addition, because the folding stage is the most time consuming stage, it makes a checkpiont for each folding process after folding every CHECKPOINT_SIZE sequences. If the pipeline is killed for some reason in the middle of folding, it can be restarted using <code>recover</code> command from where it was stopped. The default value is 3000.</li>
</ol>

<h3>c. Run the pipeline.</h3>

<p>With the configuration file ready, the pipeline can be used to predict miRNAs. The miR-PREFeR pipeline can be run as following:</p>

<pre><code>python miR_PREFeR.py [options] command configfile
</code></pre>

<p>Currently, the following options are available:</p>

<ol>
<li><strong>-h</strong>: Show help information.</li>
<li><strong>-L</strong>: Generate a log file. From the log file you can find the status during the running of the pipeline. <strong>Recommend to always use this option.</strong></li>
<li><strong>-k</strong>: After finish the whole pipeline, do not remove the temporary folder that contains the intermediate files. This will save disk space. If it's not specified, you can delete the temporary folder after getting the result.</li>
</ol>

<p><code>command</code> could be one of <code>check</code>, <code>pipeline</code>, <code>prepare</code>, <code>candidate</code>, <code>fold</code>, <code>predict</code>, and <code>recover</code>. (Run <code>python miR_PREFeR.py -h</code> to see help on options and commands.)</p>

<ol>
<li><strong>check:</strong> Check the presence of the depended programs (RNALfold and samtools), and check the recovery information (Shows which stage the previous computation on the same data was ceased. See the <code>recover</code> command in 6.).</li>
<li><strong>pipeline:</strong> Run the whole pipeline. That is, run <code>prepare</code>, <code>candidate</code>,<code>fold</code> and <code>predict</code> sequentially. This is the normal way to run miR-PREFeR.</li>
<li><strong>prepare:</strong>  Run the first step of the pipeline. This step prepares some data files needed in the following steps.</li>
<li><strong>candidate:</strong>  Identify possible candidate regions. This step can ONLY be run if the <code>prepare</code> step has been finished on the configfile file.</li>
<li><strong>fold:</strong> Fold the candidate regions. This step can ONLY be run if the <code>prepare</code> and <code>candidate</code> steps have been finished on the configfile file.</li>
<li><strong>predict:</strong> Predict miRNA loci. This step can ONLY be run if the <code>prepare</code>, <code>candidate</code> and <code>fold</code> steps have been finished on the configfile file.</li>
<li><p><strong>recover:</strong> Tries its best to recover an unfinished job and continues to run from the unfinished step it was ceased. This is designed to easily continue a job other than re-run the whole pipeline from start. For example, if a job was kill half way (This happens sometime. For example, the job takes too long time and the user killed it. Or, if the job was run on an cluster and was killed halfway because some resources exceed the max values.) In these cases, one does not need to run the pipeline from start, the job can be continued by using the <code>recover</code> command:</p>

<p><code>python miR_PREFeR.py -L recover configfile</code></p></li>
</ol>

<p><strong>NOTE on job recovery:</strong> To recover and continue a job, the intermediate output
  files in the temporary folder (See the Output section) should not be
  removed. All files needed to do recovery are in this folder.</p>

<h2>4. Output</h2>

<p>The miR-PREFeR pipeline produces several files which contains information about
the predicted miRNAs:</p>

<ol>
<li>An HTML file that contains a table which summarizes the results of the miRNA
predictions by miR-PREFeR. It contains the number of predictions, the length
distribution of the predicted mature miRNA sequences, the list of all
pre-miRNA sequences and their secondary structures, and the detailed read
mapping profiles against the pre-miRNA sequences. In addition, this table
provides search links to mirBase. An example output HTML page can be found at
<a href="http://goo.gl/YQSCP8">http://goo.gl/YQSCP8</a>.</li>
<li>A csv file that contains detailed information about each prediction. The
information in the table is the same as in the HTML file. This is provided
for easier downstream processing/analysis.</li>
<li>A list of read mapping result files in the folder <code>readmapping</code>. Each file
corresponds to an miRNA and shows the reads mapped to the precursor region.</li>
<li>A gff3 format annotation file that contains the predicted miRNAs.</li>
<li>Two fasta files that contain the mature sequences and the precursor sequences
of the predicted miRNAs.</li>
<li>A <code>.ss</code> file that contains the sequence and the predicted structure of each
miRNA.</li>
<li>A file named <code>miRNA.stat.txt</code> that shows the basic statistics of the mature
sequences.</li>
</ol>

<p>An example output folder that contains all the files can be found at
<a href="http://goo.gl/f4Jb2p">http://goo.gl/f4Jb2p</a></p>

<p>During the running of the pipeline, there is a folder named NAME_PREFIX +
"_tmp" in the output folder. The folder contains intermediate files generated by
the pipeline. If the job is finished without interruption, the folder is deleted
by default (Specify the <code>-k</code> option can keep the folder, but this is usually not
needed.).  <strong>If the job is interrupted half way, then the folder is not removed
by default, so that one can run the job in the <code>recover</code> mode to continue the
computation. (Because all files needed to do recovery are in this temp folder,
so if one wants to recover an interrupted job, please do not remove the
folder. The folder will be removed automatically once the whole pipeline is
successfully run.).</strong></p>

<h3>5. Conducting alignment using Bowtie.</h3>

<p>a1. indexing the genome sequences:</p>

<pre><code>bowtie-build -q -f TAIR10.fas bowtie-index/TAIR10
</code></pre>

<p>a2. conduct the alignment. Here are example bowtie commands for the same files processed in the previous step:</p>

<pre><code>bowtie -a -v 0 -p 8 -m 30 -S bowtie-index/TAIR10  -f SAMPLE1.fasta.processed &gt; SAMPLE1.sam 2&gt; SMAPLE1.log
bowtie -a -v 0 -p 8 -m 30 -S bowtie-index/TAIR10  -f SAMPLE2.fasta.processed &gt; SAMPLE2.sam 2&gt; SMAPLE2.log
bowtie -a -v 0 -p 8 -m 30 -S bowtie-index/TAIR10  -f SAMPLE3.fasta.processed &gt; SAMPLE3.sam 2&gt; SMAPLE3.log
</code></pre>

<p>The -v 0 option allows no mismatch in the alignment. The -m 10 option discards reads that can be mapped to more than 30 positions. The -p 8 option uses 8 threads to do the alignment. More options of the parameters can be found at <a href="http://bowtie-bio.sourceforge.net/index.shtml">http://bowtie-bio.sourceforge.net/index.shtml</a>.</p>

<p><strong>Requirement for the SAM file format:</strong></p>

<p>The pipeline requires the input SAM alignment files:</p>

<p>a.  The SAM files <strong>MUST have headers</strong>. This is a example header:</p>

<pre><code>@HD     VN:1.0  SO:unsorted
@SQ     SN:Chr1 LN:30427671
@SQ     SN:Chr2 LN:19698289
@SQ     SN:Chr3 LN:23459830
@SQ     SN:Chr4 LN:18585056
@SQ     SN:Chr5 LN:26975502
@PG     ID:Bowtie       VN:0.12.8       CL:"bowtie -a -v 1 -p 10 -S bowtie-index/TAIR10 -f ../SAMPLE1.fasta.processed"
</code></pre>

<p>You can use "<code>samtools view -H yoursamfile</code>" to check whether your SAM file has the header section. Bowtie produces headers for SAM format alignment files by default.</p>

<p>b.  The flag column (the second column of the alignment section) of the SAM file <strong>MUST be in a integer.</strong>. The flag generated by Bowtie/or generated by using "<code>samtools view</code>" command, by default, is an integer. Samtools provides options to output SAMfiles with Hex (-x option) or string (-X option) format flags. If your SAM file uses these two types of flags, you either need to generate your SAM file again, or convert it to the required format.</p>

<p>c.  The read identifier (the first column of the alignment section) must be in the required format as described in the fasta file preparation section. The <code>process-reads-fasta.py</code> and <code>convert-mirdeep2-fasta.py</code> scripts can help you generate fasta files of the right format.</p>

<p>The miR-PREFeR pipeline checks the SAM format and if the format is not correct, it prints messages about which format problems it encountered and stops analysis.</p>

</body>
</html>
